detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js

```diff
@@ -17,43 +17,43 @@ class GenyCloudDriver extends AndroidDriver {
 
     const exec = new GenyCloudExec();
     const instanceNaming = new InstanceNaming(); // TODO should consider a permissive impl for debug/dev mode. Maybe even a custom arg in package.json (Detox > ... > genycloud > sharedAccount: false)
-    this.recipeService = new RecipesService(exec, logger);
-    this.instanceLookupService = new InstanceLookupService(exec, instanceNaming, this.deviceRegistry);
-    this.instanceLifecycleService = new InstanceLifecycleService(exec, instanceNaming);
-    this.deviceQueryHelper = new DeviceQueryHelper(this.recipeService);
-    this.allocationHelper = new AllocationHelper(this.instanceLookupService, this.instanceLifecycleService);
+    this.deviceRegistry = new GenyDeviceRegistryWrapper(this.deviceRegistry);
+
+    const recipeService = new RecipesService(exec, logger);
+    const instanceLookupService = new InstanceLookupService(exec, instanceNaming, this.deviceRegistry);
+    const instanceLifecycleService = new InstanceLifecycleService(exec, instanceNaming);
+    this._deviceQueryHelper = new DeviceQueryHelper(recipeService);
+    this._deviceAllocator = new GenyCloudDeviceAllocator(this.deviceRegistry, instanceLookupService, instanceLifecycleService);
   }
 
   get name() {
     return this._name;
   }
 
   async acquireFreeDevice(deviceQuery) {
-    const recipe = await this.deviceQueryHelper.getRecipeFromQuery(deviceQuery);
-    this.assertRecipe(deviceQuery, recipe);
+    const recipe = await this._deviceQueryHelper.getRecipeFromQuery(deviceQuery);
+    this._assertRecipe(deviceQuery, recipe);
 
-    const cookie = { coldBoot: false };
-    const adbName = await this.allocateDevice({
-      recipe,
-      cookie,
-      toString: () => recipe.toString(),
-    });
+    const { instance, isNew } = await this._deviceAllocator.allocateDevice(recipe);
+    const { adbName, uuid } = instance;
 
-    await this.emitter.emit('bootDevice', { coldBoot: cookie.coldBoot, deviceId: adbName, type: recipe.name});
+    await this.emitter.emit('bootDevice', { coldBoot: isNew, deviceId: adbName, type: recipe.name});
     await this.adb.apiLevel(adbName);
     await this.adb.disableAndroidAnimations(adbName);
 
-    this._name = `GenyCloud-${adbName}`;
-    return adbName;
+    this._name = `GenyCloud:${instance.name} (${uuid} ${adbName})`;
+    return instance;
   }
 
-  async doAllocateDevice({ recipe, cookie }) {
-    const { instance, coldBooted } = await this.allocationHelper.allocateInstance(recipe.uuid);
-    cookie.coldBoot = coldBooted;
-    return instance.adbName;
```

Comment:
No self implementation, no allocation helper. Everything is now in `GenyCloudDeviceAllocator`.